File: instant-quote/src/lib/leadtime.ts
// Ship-date computation from a Europe/Warsaw wall clock with a 14:00 same-day
// cutoff, skipping weekends. No date library: we extract the Warsaw wall-clock
// via Intl, then do integer calendar math (DST-safe — we only ever step whole
// calendar days after extracting the local date).

import { PRICING, type LeadTimeId } from './pricing-config'

export interface WarsawNow {
  y: number
  m: number // 1-12
  d: number
  hour: number
  minute: number
  weekday: number // 1=Mon .. 7=Sun
}

const WARSAW = 'Europe/Warsaw'

const partsFormatter = new Intl.DateTimeFormat('en-GB', {
  timeZone: WARSAW,
  year: 'numeric',
  month: '2-digit',
  day: '2-digit',
  hour: '2-digit',
  minute: '2-digit',
  hourCycle: 'h23',
  weekday: 'short',
})

const WEEKDAY_INDEX: Record<string, number> = {
  Mon: 1,
  Tue: 2,
  Wed: 3,
  Thu: 4,
  Fri: 5,
  Sat: 6,
  Sun: 7,
}

export function getWarsawNow(now: Date): WarsawNow {
  const parts = partsFormatter.formatToParts(now)
  const get = (t: string) => parts.find((p) => p.type === t)?.value ?? ''
  return {
    y: Number(get('year')),
    m: Number(get('month')),
    d: Number(get('day')),
    hour: Number(get('hour')),
    minute: Number(get('minute')),
    weekday: WEEKDAY_INDEX[get('weekday')] ?? 1,
  }
}

interface CalDate {
  y: number
  m: number
  d: number
}

// Weekday (1=Mon..7=Sun) for a calendar date, via UTC (no TZ involved).
function weekdayOf({ y, m, d }: CalDate): number {
  const dow = new Date(Date.UTC(y, m - 1, d)).getUTCDay() // 0=Sun..6=Sat
  return dow === 0 ? 7 : dow
}

function isWeekend(date: CalDate): boolean {
  const wd = weekdayOf(date)
  return wd === 6 || wd === 7
}

function addCalendarDays(date: CalDate, n: number): CalDate {
  const dt = new Date(Date.UTC(date.y, date.m - 1, date.d + n))
  return {
    y: dt.getUTCFullYear(),
    m: dt.getUTCMonth() + 1,
    d: dt.getUTCDate(),
  }
}

function nextBusinessDay(date: CalDate): CalDate {
  let next = addCalendarDays(date, 1)
  while (isWeekend(next)) next = addCalendarDays(next, 1)
  return next
}

function addBusinessDays(date: CalDate, n: number): CalDate {
  let result = date
  for (let i = 0; i < n; i++) result = nextBusinessDay(result)
  return result
}

export interface ShipDate {
  date: CalDate
  dispatchStartsToday: boolean // false if weekend or past 14:00 cutoff
  label: string // e.g. "Thu, 16 Jul"
}

const labelFormatter = new Intl.DateTimeFormat('en-GB', {
  weekday: 'short',
  day: 'numeric',
  month: 'short',
})

function formatLabel({ y, m, d }: CalDate): string {
  return labelFormatter.format(new Date(Date.UTC(y, m - 1, d)))
}

export function computeShipDate(
  lead: LeadTimeId,
  now: Date = new Date(),
): ShipDate {
  const w = getWarsawNow(now)
  const today: CalDate = { y: w.y, m: w.m, d: w.d }

  const todayIsBusiness = !isWeekend(today)
  const beforeCutoff = w.hour < PRICING.sameDayCutoffHour
  const dispatchStartsToday = todayIsBusiness && beforeCutoff

  // Day 0 for the lead-time countdown: today if we can still dispatch today,
  // otherwise the next business day.
  const day0 = dispatchStartsToday ? today : nextBusinessDay(today)

  const businessDays = PRICING.leadTimes[lead].businessDays
  const shipDate = addBusinessDays(day0, businessDays)

  return {
    date: shipDate,
    dispatchStartsToday,
    label: formatLabel(shipDate),
  }
}

export function formatWarsawClock(now: Date = new Date()): string {
  const w = getWarsawNow(now)
  const hh = String(w.hour).padStart(2, '0')
  const mm = String(w.minute).padStart(2, '0')
  return `${hh}:${mm}`
}

File: backend/internal/leadtime/leadtime.go
// Package leadtime ports src/lib/leadtime.ts: ship-date computation from a
// Europe/Warsaw wall clock with a 14:00 same-day cutoff, skipping weekends
// (no public-holiday handling). Calendar math steps whole days after
// extracting the Warsaw local date, so DST cannot skew it.
package leadtime

import (
	"fmt"
	"time"
)

var warsaw = mustLoadWarsaw()

func mustLoadWarsaw() *time.Location {
	loc, err := time.LoadLocation("Europe/Warsaw")
	if err != nil {
		panic(err) // tzdata is embedded via time/tzdata in cmd/api
	}
	return loc
}

type CalDate struct {
	Y int `json:"y"`
	M int `json:"m"` // 1-12
	D int `json:"d"`
}

type ShipDate struct {
	Date CalDate `json:"date"`
	// False if weekend or past the 14:00 cutoff.
	DispatchStartsToday bool   `json:"dispatchStartsToday"`
	Label               string `json:"label"` // e.g. "Thu 16 Jul"
}

// warsawDate extracts the Warsaw wall-clock date and hour for an instant.
func warsawDate(now time.Time) (CalDate, int) {
	w := now.In(warsaw)
	return CalDate{Y: w.Year(), M: int(w.Month()), D: w.Day()}, w.Hour()
}

// weekdayOf: 1=Mon .. 7=Sun for a calendar date (no TZ involved).
func weekdayOf(d CalDate) int {
	wd := time.Date(d.Y, time.Month(d.M), d.D, 0, 0, 0, 0, time.UTC).Weekday()
	if wd == time.Sunday {
		return 7
	}
	return int(wd)
}

func isWeekend(d CalDate) bool {
	wd := weekdayOf(d)
	return wd == 6 || wd == 7
}

func addCalendarDays(d CalDate, n int) CalDate {
	t := time.Date(d.Y, time.Month(d.M), d.D+n, 0, 0, 0, 0, time.UTC)
	return CalDate{Y: t.Year(), M: int(t.Month()), D: t.Day()}
}

func nextBusinessDay(d CalDate) CalDate {
	next := addCalendarDays(d, 1)
	for isWeekend(next) {
		next = addCalendarDays(next, 1)
	}
	return next
}

func addBusinessDays(d CalDate, n int) CalDate {
	result := d
	for i := 0; i < n; i++ {
		result = nextBusinessDay(result)
	}
	return result
}

// Short names match Chrome's Intl en-GB output, which the app displayed
// before the port (note "Sept", not "Sep").
var weekdayShort = [8]string{"", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"}
var monthShort = [13]string{"", "Jan", "Feb", "Mar", "Apr", "May", "Jun",
	"Jul", "Aug", "Sept", "Oct", "Nov", "Dec"}

func formatLabel(d CalDate) string {
	return fmt.Sprintf("%s %d %s", weekdayShort[weekdayOf(d)], d.D, monthShort[d.M])
}

// ComputeShipDate returns the estimated ship date for a lead time of
// businessDays, given the same-day dispatch cutoff hour (Warsaw wall clock).
func ComputeShipDate(businessDays, cutoffHour int, now time.Time) ShipDate {
	today, hour := warsawDate(now)

	todayIsBusiness := !isWeekend(today)
	beforeCutoff := hour < cutoffHour
	dispatchStartsToday := todayIsBusiness && beforeCutoff

	// Day 0 for the lead-time countdown: today if we can still dispatch
	// today, otherwise the next business day.
	day0 := today
	if !dispatchStartsToday {
		day0 = nextBusinessDay(today)
	}

	shipDate := addBusinessDays(day0, businessDays)

	return ShipDate{
		Date:                shipDate,
		DispatchStartsToday: dispatchStartsToday,
		Label:               formatLabel(shipDate),
	}
}
